### PR TITLE
[Femdem] Improving the FemDem elements in 3D for the remeshing

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm adding some improvements in the femdem files in order to perform the remeshing.
Besides I've added a minor modification of the Mohr-Coulomb yield surface in structural mechanics App.

### CODE REVIEW

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp [hunk 1]
```diff
@@ -124,7 +124,7 @@ void FemDem3DElement::InitializeInternalVariablesAfterMapping()
 
 void FemDem3DElement::ComputeEdgeNeighbours(ProcessInfo &rCurrentProcessInfo)
 {
-	std::vector<std::vector<Element *>> EdgeNeighboursContainer;
+	std::vector<std::vector<Element *>> p_edge_neighboursContainer;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_element.cpp]: Technically the p is inside the vector, so I am not sure about the name @philbucher ?

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp [hunk 2]
```diff
@@ -149,7 +149,7 @@ void FemDem3DElement::ComputeEdgeNeighbours(ProcessInfo &rCurrentProcessInfo)
 	Matrix nodes_indexes = ZeroMatrix(6, 2);
 	this->SetNodeIndexes(nodes_indexes);
 
-	// Loop over EDGES to assign the elements that share that edge -> Fill mEdgeNeighboursContainer
+	// Loop over EDGES to assign the elements that share that edge -> Fill mp_edge_neighboursContainer
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_element.cpp]: Member variables are CammelCase, it was fine

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp [hunk 3]
```diff
@@ -902,7 +901,7 @@ void FemDem3DElement::CalculateOnIntegrationPoints(
 	if (rVariable == DAMAGE_ELEMENT) {
 		rOutput.resize(1);
 		for (unsigned int PointNumber = 0; PointNumber < 1; PointNumber++) {
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_element.cpp]: In this case point_number

File: applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp [hunk 4]
```diff
@@ -468,5 +479,210 @@ void FemDem3DLargeDisplacementElement::CalculateAndAddInternalForcesVector(
     noalias(rRightHandSideVector) -= IntegrationWeight * prod(trans(rB), rStressVector);
 }
 
+// Methods to compute the tangent tensor by numerical derivation
+void FemDem3DLargeDisplacementElement::CalculateTangentTensor(
+	Matrix& rTangentTensor,
+	const Vector& rStrainVectorGP,
+	const Vector& rStressVectorGP,
+    const Matrix& rDeformationGradientGP,
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp:487]: Check this tabulations, we use 4 spaces, there are more

File: applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp [hunk 5]
```diff
@@ -468,5 +479,210 @@ void FemDem3DLargeDisplacementElement::CalculateAndAddInternalForcesVector(
...
+	const Matrix& rElasticMatrix
+	)
+{
+	const double number_components = rStrainVectorGP.size();
+	rTangentTensor.resize(number_components, number_components);
+	Vector perturbed_stress, perturbed_strain;
+	perturbed_strain.resize(number_components);
+	perturbed_stress.resize(number_components);
+    Matrix perturbed_deformation_gradient;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp:496]: Like here

File: applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp [hunk 6]
```diff
@@ -468,5 +479,210 @@ void FemDem3DLargeDisplacementElement::CalculateAndAddInternalForcesVector(
...
+}
+
+
+int FemDem3DLargeDisplacementElement::CalculateVoigtIndex(
+    const SizeType VoigtSize,
+    const int ComponentI,
+    const int ComponentJ
+    )
+{
+    if (VoigtSize == 6) {
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp:547]: I think this is in MathUtils

File: applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.hpp [hunk 7]
```diff
@@ -51,39 +51,61 @@ class FemDem3DLargeDisplacementElement : public FemDem3DElement
...
-    void CalculateStressVectorPredictor(Vector& rStressVector, 
-                                        const Matrix& rConstitutiveMAtrix, 
-                                        const Vector& rStrainVector);
-
-    void CalculateAndAddInternalForcesVector(Vector& rRightHandSideVector, 
-                                            const Matrix& rB, 
-                                            const Vector& rStressVector, 
-                                            const double IntegrationWeight);
+  int GetStrainSize(){ return 6; }
+  double CalculateDerivativesOnReferenceConfiguration(Matrix& rJ0,
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.hpp:56]: Add some descriptions

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Look my comments

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp (hunk 2 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem3d_element.cpp]: done

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp (hunk 1 above)
Comment (philbucher) [reply applications/FemToDemApplication/custom_elements/femdem3d_element.cpp]: I guess I would call it `edge_neighbours_container`

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem3d_element.cpp]: done

File: applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp (hunk 6 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp:547]: I see it in the tangent tensor calculations :)

File: applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp (hunk 4 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_elements/femdem3d_large_displacement_element.cpp:487]: should be okkey now

File: applications/FemToDemApplication/custom_elements/femdem3d_element.cpp [hunk 8]
```diff
@@ -124,54 +124,54 @@ void FemDem3DElement::InitializeInternalVariablesAfterMapping()
...
-	std::vector<std::vector<Element *>> EdgeNeighboursContainer;
-	Geometry<Node<3>> &NodesCurrentElement = this->GetGeometry();
+	std::vector<std::vector<Element*>> edge_neighboursContainer;
+	Geometry<Node<3>>& r_nodes_current_element = this->GetGeometry();
 
-	Node<3> &pNode0 = NodesCurrentElement[0];
-	Node<3> &pNode1 = NodesCurrentElement[1];
-	Node<3> &pNode2 = NodesCurrentElement[2];
-	Node<3> &pNode3 = NodesCurrentElement[3];
+	Node<3> &pNode0 = r_nodes_current_element[0];
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_elements/femdem3d_element.cpp:130]: Tecnically r_node :P

**Review by loumalouomega — CHANGES_REQUESTED** [review]

**Review by loumalouomega — DISMISSED** [review]
I will approve, but still some formatting missing

**Comment by AlejandroCornejo** [comment]
srry @loumalouomega can you reapprove? :)

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**